KIPs/kip-114.md

```diff
@@ -0,0 +1,324 @@
+---
+kip: 114
+title: Signature-based random in block headers
+author: Ian (@ian0371), Ollie (@blukat29), Joseph (@jiseongnoh), and Aidan (@aidan-kwon)
+discussions-to: https://github.com/klaytn/kips/issues/114
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-05-31
+---
+
+## Simple Summary
+
+New fields in the block header as the source of on-chain random
+
+## Abstract
+
+This standard outlines an approach for generating an on-chain random value based on digital signature. Specifically, it introduces new fields, namely `randomReveal` and `mixHash`, in the block header; `randomReveal` is a verifiable random number, and `mixHash` is a derivation from a mixture of `randomReveals` and additional information. They serve as an unpredictable, verifiable, and dependable source of on-chain randomness.
+
+## Motivation
+
+An unpredictable, verifiable, and dependable on-chain randomness can be utilized in various use-cases including the protocol and applications.
+
+- Unpredictability indicates that the random should be as difficult to predict as possible.
+
+- Verifiability ensures that the random is authentic and cannot be manipulated or forged.
+
+- Dependability implies that the random is generated without trusting any external sources of randomness, such as oracles.
+
+An example of such use-cases is the block proposer selection. An exemplary requirement can be that the next block's proposer remains undisclosed until the latest block has been finalized, or that the proposer cannot tamper with deciding the next block's proposer.
+
+In this proposal, new header fields `randomReveal` and `mixHash` are introduced. `randomReveal` is the signature of the proposer, and `randomReveal`s are mixed to compute `mixHash`, which users can use for a part of the random source. `mixHash` is unpredictable, verifiable, and dependable:
+
+- (Unpredictability) the signature is random and so is `mixHash`.
+- (Verifiability) the signature can be verified with the public key of the proposer, and `mixHash` is computed in a deterministic manner.
+- (Dependability) `mixHash` only relies on the validators.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+The terms hash_to_point (hereinafter referred to as hash), public key, proof-of-possession (hereinafter referred to as pop) are from the ciphersuite [BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP\_](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-4.2.3).
+
+### Parameters
+
+| Constant     | Value |
+| ------------ | ----- |
+| `FORK_BLOCK` | TBD   |
+
+### Block Structure
+
+Beginning with `FORK_BLOCK`, client software must set the fields in a block header:
```

Comment:
Can we show the whole data structure of the block header?

Reply:
What is the purpose? If we are showing the whole header, I'd like to show both headers before and after FORK_BLOCK.

Reply:
The purpose is to give a clear view to the users. 

- If we add more fields in the future, the order could be changed.
- What are the preceding and following fields in the blocks?
- How is the RLP encoding order?

What about showing like this instead of showing before and after?

```
type Block {
 OldField1 int
 OldField2 int
 NewField1 int // newly added
}
```

Reply:
> If we add more fields in the future, the order could be changed.
> What are the preceding and following fields in the blocks?

I thought altering header fields could be done only via "append", and switching orders is impossible.

Anyways, I think there's no harm in showing the whole header. I'll add it like the suggestion.

> How is the RLP encoding order?

The order is described (15th, 16th field).

Reply:
added in [4e6f40c](https://github.com/klaytn/kips/pull/144/commits/4e6f40cd4bc7e233396f87d25be2ee6b2ec674b8)